[TOPIC-GPIO] drivers: gpio_sifive: Update for new GPIO API

nategraff-sifive opened the pull request:
Update the SiFive GPIO driver for the new GPIO API, including replacing deprecated flag names, implementing the `port_*` API surface, and moving interrupt configuration to `pin_interrupt_configure`.

One of the drivers tracked in #18530

nategraff-sifive committed:
drivers: gpio_sifive: Update for new GPIO API

Update the Sifive GPIO driver for new API

Signed-off-by: Nathaniel Graff <nathaniel.graff@sifive.com>

carlescufi commented:
@nategraff-sifive 
The topic branch has been rebased. Could you please rebase this PR against it?

nategraff-sifive commented:
@carlescufi rebased 👍

mnkp commented on drivers/gpio/gpio_sifive.c:
```diff
@@ -117,25 +117,34 @@ static int gpio_sifive_config(struct device *dev,
-	if (flags & GPIO_DIR_OUT) {
+	if (flags & GPIO_OUTPUT) {
 		gpio->in_en &= ~BIT(pin);
 		gpio->out_en |= BIT(pin);
```
It seems like on this platform input and output can be enabled / disabled independently. If this is the case we should write the code as follows
```
	if ((flags & GPIO_OUTPUT) != 0) {
￼		gpio->out_en |= BIT(pin);
	} else {
￼		gpio->out_en &= ~BIT(pin);
	}
```
or alternatively use `WRITE_BIT` macro
```
	WRITE_BIT(gpio->out_en, pin, flags & GPIO_OUTPUT);
```
We should do the same for input flag.

mnkp commented on drivers/gpio/gpio_sifive.c:
```diff
@@ -117,25 +117,34 @@ static int gpio_sifive_config(struct device *dev,
 			gpio->invert |= BIT(pin);
 		} else {
 			gpio->invert &= ~BIT(pin);
 		}
```
Converting between logical and physical value is handled entirely in `include/drivers/gpio.h`. We should remove it from the driver.

mnkp commented on drivers/gpio/gpio_sifive.c:
```diff
@@ -117,25 +117,34 @@ static int gpio_sifive_config(struct device *dev,
+			gpio->out_val &= ~BIT(pin);
+		} else if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			gpio->out_val |= BIT(pin);
+		}
```
Initial value should be set before we configure pin as an output. Otherwise user may observe a glitch.

mnkp commented on drivers/gpio/gpio_sifive.c:
```diff
@@ -117,25 +117,34 @@ static int gpio_sifive_config(struct device *dev,
-		if (flags & GPIO_POL_INV) {
+		if (flags & GPIO_ACTIVE_LOW) {
 			return -EINVAL;
 		}
```
Converting between logical and physical value is handled entirely in `include/drivers/gpio.h`. We should remove it from the driver.

mnkp commented on drivers/gpio/gpio_sifive.c:
```diff
@@ -152,56 +161,8 @@ static int gpio_sifive_config(struct device *dev,
 		} else {
 			gpio->pue &= ~BIT(pin);
```
```
		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_UP) {
```
test should be written as
```
		if ((flags & GPIO_PULL_UP) != 0) {
```
The `GPIO_PUD_` macros will be deprecated.

mnkp commented on drivers/gpio/gpio_sifive.c:
```diff
@@ -117,25 +117,34 @@ static int gpio_sifive_config(struct device *dev,
 	}
 
```
The driver should return `-ENOTSUP` if an open source / open drain configuration is requested.
```
	if ((flags & GPIO_SINGLE_ENDED) != 0) {
		return -ENOTSUP;
	}
```

mnkp commented on drivers/gpio/gpio_sifive.c:
```diff
@@ -284,6 +245,108 @@ static int gpio_sifive_read(struct device *dev,
+			gpio->low_ie |= BIT(pin);
+		} else {
+			return -ENOTSUP;
+		}
```
I'll add an assert to `include/drivers/gpio.h` to guard for this condition. Let's remove it from the driver.

nategraff-sifive commented:
@mnkp, thanks for your feedback! I've updated the PR based on your suggestions. What do you think of the new configure? I'm setting/clearing all of the bits with the `WRITE_BIT` macro now.

mnkp commented on drivers/gpio/gpio_sifive.c:
```diff
@@ -116,93 +116,26 @@ static int gpio_sifive_config(struct device *dev,
-	}
+	/* Set the initial output value before enabling output to avoid
+	 * glitches */
+	WRITE_BIT(gpio->out_val, pin, flags & GPIO_OUTPUT_INIT_HIGH);
```
This part we need to write differently since the API supports a case where the output value is not initialized before enabling it. One way to do it would be
```
if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
	gpio->out_val |= BIT(pin);
}
if ((flags & GPIO_OUTPUT_INIT_LOW) != 0) {
	gpio->out_val &= ~BIT(pin);
}
```
I would also move this code section below line 134. Then pull up / pull down configuration and output / input would be next to each other.

mnkp reviewed (commented):
Thanks! LGTM with one remark.

It would be great if you could verify that the driver is passing `tests/drivers/gpio/gpio_api_1pin/` testcase. You need to run it on the board which has `LED0` defined in the board DTS file.

nategraff-sifive commented:
I'm getting some test failures on `tests/drivers/gpio/gpio_api_1pin/` on `hifive1_revb`

```
***** Booting Zephyr OS build zephyr-v2.0.0-1178-gfb2acd99d9d8 *****
Running test suite gpio_api_1pin_test
===================================================================
starting test - test_gpio_pin_configure_push_pull
Running test on port=gpio_0, pin=19
PASS - test_gpio_pin_configure_push_pull
===================================================================
starting test - test_gpio_pin_configure_single_ended
Running test on port=gpio_0, pin=19
Pull Up / Pull Down pin bias is not supported
SKIP - test_gpio_pin_configure_single_ended
===================================================================
starting test - test_gpio_pin_set_get_raw
Running test on port=gpio_0, pin=19
PASS - test_gpio_pin_set_get_raw
===================================================================
starting test - test_gpio_pin_set_get
Running test on port=gpio_0, pin=19
PASS - test_gpio_pin_set_get
===================================================================
starting test - test_gpio_pin_set_get_active_high
Running test on port=gpio_0, pin=19
Step 1: Set logical, get logical and physical pin value
Step 2: Set physical, get logical and physical pin value
PASS - test_gpio_pin_set_get_active_high
===================================================================
starting test - test_gpio_pin_set_get_active_low
Running test on port=gpio_0, pin=19
Step 1: Set logical, get logical and physical pin value
Step 2: Set physical, get logical and physical pin value
PASS - test_gpio_pin_set_get_active_low
===================================================================
starting test - test_gpio_pin_toggle
Running test on port=gpio_0, pin=19
PASS - test_gpio_pin_toggle
===================================================================
starting test - test_gpio_port_set_masked_get_raw
Running test on port=gpio_0, pin=19
PASS - test_gpio_port_set_masked_get_raw
===================================================================
starting test - test_gpio_port_set_masked_get
Running test on port=gpio_0, pin=19
PASS - test_gpio_port_set_masked_get
===================================================================
starting test - test_gpio_port_set_masked_get_active_high
Running test on port=gpio_0, pin=19
Step 1: Set logical, get logical and physical port value
Step 2: Set physical, get logical and physical port value
PASS - test_gpio_port_set_masked_get_active_high
===================================================================
starting test - test_gpio_port_set_masked_get_active_low
Running test on port=gpio_0, pin=19
Step 1: Set logical, get logical and physical port value
Step 2: Set physical, get logical and physical port value
PASS - test_gpio_port_set_masked_get_active_low
===================================================================
starting test - test_gpio_port_set_bits_clear_bits_raw
Running test on port=gpio_0, pin=19
PASS - test_gpio_port_set_bits_clear_bits_raw
===================================================================
starting test - test_gpio_port_set_bits_clear_bits
Running test on port=gpio_0, pin=19
PASS - test_gpio_port_set_bits_clear_bits
===================================================================
starting test - test_gpio_port_set_clr_bits_raw
Running test on port=gpio_0, pin=19
PASS - test_gpio_port_set_clr_bits_raw
===================================================================
starting test - test_gpio_port_set_clr_bits
Running test on port=gpio_0, pin=19
PASS - test_gpio_port_set_clr_bits
===================================================================
starting test - test_gpio_port_toggle
Running test on port=gpio_0, pin=19
PASS - test_gpio_port_toggle
===================================================================
starting test - test_gpio_int_edge_rising
Running test on port=gpio_0, pin=19

    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_api_1pin/src/test_pin_interrupt.c:105: test_gpio_pin_interrupt_edge: (cb_count not equal to cb_count_expected)
Test point 0: Pin interrupt triggered invalid number of times on rising/to active edge
FAIL - test_gpio_int_edge_rising
===================================================================
starting test - test_gpio_int_edge_falling
Running test on port=gpio_0, pin=19

    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_api_1pin/src/test_pin_interrupt.c:113: test_gpio_pin_interrupt_edge: (cb_count not equal to cb_count_expected)
Test point 0: Pin interrupt triggered invalid number of times on falling/to inactive edge
FAIL - test_gpio_int_edge_falling
===================================================================
starting test - test_gpio_int_edge_both
Running test on port=gpio_0, pin=19

    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_api_1pin/src/test_pin_interrupt.c:105: test_gpio_pin_interrupt_edge: (cb_count not equal to cb_count_expected)
Test point 0: Pin interrupt triggered invalid number of times on rising/to active edge
FAIL - test_gpio_int_edge_both
===================================================================
starting test - test_gpio_int_edge_to_active
Step 1: Configure pin as active high
Running test on port=gpio_0, pin=19

    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_api_1pin/src/test_pin_interrupt.c:105: test_gpio_pin_interrupt_edge: (cb_count not equal to cb_count_expected)
Test point 0: Pin interrupt triggered invalid number of times on rising/to active edge
FAIL - test_gpio_int_edge_to_active
===================================================================
starting test - test_gpio_int_edge_to_inactive
Step 1: Configure pin as active high
Running test on port=gpio_0, pin=19

    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_api_1pin/src/test_pin_interrupt.c:113: test_gpio_pin_interrupt_edge: (cb_count not equal to cb_count_expected)
Test point 0: Pin interrupt triggered invalid number of times on falling/to inactive edge
FAIL - test_gpio_int_edge_to_inactive
===================================================================
starting test - test_gpio_int_level_high
Running test on port=gpio_0, pin=19

    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_api_1pin/src/test_pin_interrupt.c:186: test_gpio_pin_interrupt_level: (cb_count not equal to cb_count_expected)
Test point 0: Pin interrupt triggered invalid number of times on level 1
FAIL - test_gpio_int_level_high
===================================================================
starting test - test_gpio_int_level_low
Running test on port=gpio_0, pin=19

    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_api_1pin/src/test_pin_interrupt.c:186: test_gpio_pin_interrupt_level: (cb_count not equal to cb_count_expected)
Test point 0: Pin interrupt triggered invalid number of times on level 0
FAIL - test_gpio_int_level_low
===================================================================
starting test - test_gpio_int_level_active
Step 1: Configure pin as active high
Running test on port=gpio_0, pin=19

    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_api_1pin/src/test_pin_interrupt.c:186: test_gpio_pin_interrupt_level: (cb_count not equal to cb_count_expected)
Test point 0: Pin interrupt triggered invalid number of times on level 1
FAIL - test_gpio_int_level_active
===================================================================
starting test - test_gpio_int_level_inactive
Step 1: Configure pin as active high
Running test on port=gpio_0, pin=19

    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_api_1pin/src/test_pin_interrupt.c:186: test_gpio_pin_interrupt_level: (cb_count not equal to cb_count_expected)
Test point 0: Pin interrupt triggered invalid number of times on level 0
FAIL - test_gpio_int_level_inactive
===================================================================
starting test - test_gpio_pin_toggle_visual
Running test on port=gpio_0, pin=19
LED ON
LED OFF
LED ON
LED OFF
PASS - test_gpio_pin_toggle_visual
===================================================================
Test suite gpio_api_1pin_test failed.
===================================================================
PROJECT EXECUTION FAILED
```

nategraff-sifive commented:
One other thing I noticed while running the test is that the LED polarity is inverted from what the test thinks. The LED0 GPIO pin is on the cathode of the LED, so when the GPIO is driven low the LED is turned on. During test_gpio_pin_toggle_visual, the LED is ON when the test says it's OFF.

[Schematic for the HiFive1 Rev B LEDs (Go to page 3)](https://sifive.cdn.prismic.io/sifive%2Fa4546ced-0922-4d87-9334-e97c1a9fd9a5_hifive1.b01.schematics.pdf)

What are the odds this is the cause of the interrupt failures in the test?

mnkp commented:
According to the testcase log the interrupts are not triggered at all. Since test for `GPIO_INT_EDGE_BOTH` also fails, so it's not polarity related. Does the button sample `samples/basic/button/` work?

The trace messages printed by the testcase are not very good at the moment since they don't print actual and expected values. I didn't do it explicitly in the testcase hoping that this issue will be solved globally. The assert macros should print actual and expected values automatically, always. Unfortunately, that's not so easy to implement.

mnkp commented:
To fix the LED polarity you need to add `GPIO_ACTIVE_LOW` to the board dts file `boards/riscv/hifive1_revb/hifive1_revb.dts`. Look e.g. at `boards/arm/sam_e70_xplained/sam_e70_xplained.dts` as an example.

mnkp commented:
There is one more issue I overlooked, in the `gpio_sifive_irq_handler` function the interrupt status has to be cleared first, before `gpio_fire_callbacks` function is called. Otherwise a second edge interrupt that happens when the callback is being executed will be missed.

mnkp commented:
I will re-review the code when the pending issues are solved.

carlescufi commented:
@nategraff-sifive could you address the remaining issues please?

nategraff-sifive commented:
@carlescufi I'm working on it!

carlescufi commented:
> @carlescufi I'm working on it!

Thanks Nate.

nategraff-sifive commented:
Fixed the polarity and made the change you requested to the irq_handler. I'm still tracking down the test failures though.

nategraff-sifive commented:
Looking at `build/zephyr/include/generated/generated_dts_board_unfixed.h`, the `DT_SIFIVE_GPIO_10012000_IRQ_*` macros look wrong to me:

```
/*  Devicetree node: /soc/gpio@10012000  */
/*  Binding (compatible = sifive,gpio0): $ZEPHYR_BASE/dts/bindings/gpio/sifive,gpio0.yaml  */
/*  Binding description: This is a representation of the SiFive GPIO nodes  */
#define DT_SIFIVE_GPIO0_10012000_BASE_ADDRESS       0x10012000
#define DT_INST_0_SIFIVE_GPIO0_BASE_ADDRESS         DT_SIFIVE_GPIO0_10012000_BASE_ADDRESS
#define DT_SIFIVE_GPIO0_10012000_CONTROL_BASE_ADDRESS DT_SIFIVE_GPIO0_10012000_BASE_ADDRESS
#define DT_INST_0_SIFIVE_GPIO0_CONTROL_BASE_ADDRESS DT_SIFIVE_GPIO0_10012000_BASE_ADDRESS
#define DT_SIFIVE_GPIO0_10012000_SIZE               4096
#define DT_INST_0_SIFIVE_GPIO0_SIZE                 DT_SIFIVE_GPIO0_10012000_SIZE
#define DT_SIFIVE_GPIO0_10012000_CONTROL_SIZE       DT_SIFIVE_GPIO0_10012000_SIZE
#define DT_INST_0_SIFIVE_GPIO0_CONTROL_SIZE         DT_SIFIVE_GPIO0_10012000_SIZE
#define DT_SIFIVE_GPIO0_10012000_IRQ_0              2315
#define DT_INST_0_SIFIVE_GPIO0_IRQ_0                DT_SIFIVE_GPIO0_10012000_IRQ_0
#define DT_SIFIVE_GPIO0_10012000_IRQ_1              2571
#define DT_INST_0_SIFIVE_GPIO0_IRQ_1                DT_SIFIVE_GPIO0_10012000_IRQ_1
#define DT_SIFIVE_GPIO0_10012000_IRQ_2              2827
#define DT_INST_0_SIFIVE_GPIO0_IRQ_2                DT_SIFIVE_GPIO0_10012000_IRQ_2
#define DT_SIFIVE_GPIO0_10012000_IRQ_3              3083
#define DT_INST_0_SIFIVE_GPIO0_IRQ_3                DT_SIFIVE_GPIO0_10012000_IRQ_3
#define DT_SIFIVE_GPIO0_10012000_IRQ_4              3339
#define DT_INST_0_SIFIVE_GPIO0_IRQ_4                DT_SIFIVE_GPIO0_10012000_IRQ_4
#define DT_SIFIVE_GPIO0_10012000_IRQ_5              3595
#define DT_INST_0_SIFIVE_GPIO0_IRQ_5                DT_SIFIVE_GPIO0_10012000_IRQ_5
#define DT_SIFIVE_GPIO0_10012000_IRQ_6              3851
#define DT_INST_0_SIFIVE_GPIO0_IRQ_6                DT_SIFIVE_GPIO0_10012000_IRQ_6
#define DT_SIFIVE_GPIO0_10012000_IRQ_7              4107
```

Shouldn't those have the IRQ numbers (8-39) instead of these large, 4-digit numbers?

mnkp commented:
> Shouldn't those have the IRQ numbers (8-39) instead of these large, 4-digit numbers?

Yes, those defines look very wrong.

nategraff-sifive commented:
Good to know, so a (the?) problem is that the devicetree->configuration is corrupting up the IRQ line numbers for the HiFive1 and HiFive1 Rev B.

I'm very out of my depth with Zephyr's devicetree parser, especially since I get the sense that it's changed a lot since I last interacted with it. Maybe someone can put me in touch with someone who knows how that system works?

mnkp commented:
It would be probably best if you ask a question on #devicetree channel in Slack. 

Just to double check, you could run a quick test and modify `IRQ_CONNECT` macro in `gpio_sifive.c` to something like.
```
-IRQ_CONNECT(DT_INST_0_SIFIVE_GPIO0_IRQ_##n,    \
+IRQ_CONNECT((8 + n),   \
```
and verify that after the modification the interrupts are firing correctly.

nategraff-sifive commented:
The devicetree channel is a good idea, I'll ask there.

So the PLIC IRQ numbers are actually offset by 12, so the change to IRQ_CONNECT is
```
-IRQ_CONNECT(DT_INST_0_SIFIVE_GPIO0_IRQ_##n,    \
+IRQ_CONNECT((12 + 8 + n),   \
```
...which doesn't work, even though the generated `_sw_isr_table` has the SiFive GPIO irq_handlers at the right offsets 20-51.

nategraff-sifive commented:
The weird IRQ numbers are said to be a result of multi-level IRQ handling, so the issue still lies elsewhere.

nategraff-sifive commented:
I found it!

`gpio_sifive_enable_callback` calls `irq_enable(cfg->gpio_irq_base + pin);`, which no longer works because of the multi-level IRQ numbers.

mnkp commented:
Nice! Good job! :)

But it reminds me that there is one more thing to fix. The `gpio_sifive_pin_interrupt_configure` function needs to enable / disable interrupt line as instructed by the `mode` parameter. So it needs to call `irq_enable` / `irq_disable`. The `gpio_sifive_enable_callback` / `gpio_sifive_disable_callback` functions will be deprecated.

nategraff-sifive commented:
Progress update:

I just pushed a new commit with my work so far to update the driver for multi-level interrupts.

It still fails all of the interrupt tests in `tests/drivers/gpio/gpio_api_1pin`, except it gets to step 2 on `test_gpio_int_edge_to_active`
```
===================================================================
starting test - test_gpio_int_edge_to_active
Step 1: Configure pin as active high
Running test on port=gpio_0, pin=19
Step 2: Configure pin as active low
Running test on port=gpio_0, pin=19

    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_api_1pin/src/test_pin_interrupt.c:105: test_gpio_pin_interrupt_edge: (cb_count not equal to cb_count_expected)
Test point 0: Pin interrupt triggered invalid number of times on rising/to active edge
FAIL - test_gpio_int_edge_to_active
```

mnkp commented on drivers/gpio/gpio_sifive.c:
```diff
@@ -284,6 +269,110 @@ static int gpio_sifive_read(struct device *dev,
+	switch (mode) {
+	case GPIO_INT_MODE_DISABLED:
+		gpio->rise_ie &= ~BIT(pin);
+		gpio->rise_ie &= ~BIT(pin);
```
Repeated line

mnkp commented on drivers/gpio/gpio_sifive.c:
```diff
@@ -284,6 +269,110 @@ static int gpio_sifive_read(struct device *dev,
+		break;
+	case GPIO_INT_MODE_LEVEL:
+		gpio->rise_ie &= ~BIT(pin);
+		gpio->rise_ie &= ~BIT(pin);
```
Repeated line

mnkp commented on drivers/gpio/gpio_sifive.c:
```diff
@@ -284,6 +269,110 @@ static int gpio_sifive_read(struct device *dev,
+			gpio->high_ie &= ~BIT(pin);
+			gpio->low_ie |= BIT(pin);
+		}
+		irq_enable(gpio_sifive_pin_irq(cfg->gpio_irq_base, pin));
```
I don't know this platform but we probably need to clear interrupt status register before enabling the interrupt source. E.g. in the irq handler we have the following code
```
	if (gpio->rise_ip & BIT(pin)) {
		gpio->rise_ip = BIT(pin);
	} else if (gpio->fall_ip & BIT(pin)) {
		gpio->fall_ip = BIT(pin);
	} else if (gpio->high_ip & BIT(pin)) {
		gpio->high_ip = BIT(pin);
	} else if (gpio->low_ip & BIT(pin)) {
		gpio->low_ip = BIT(pin);
	}
```
On most of the hardware having 'dirty' interrupt status register would result in an interrupt being triggered immediately.

carlescufi commented:
@nategraff-sifive could you fix the CI issues and reply to @mnkp 's comment above about the interrupt handling?

nategraff-sifive commented:
Clearing the pending register before enabling the GPIO interrupt didn't fix it. I'm still digging around. The problem is definitely spurious interrupts, I can confirm that the callbacks are firing too much, not too little.

galak commented:
We rebased the topic-gpio branch against master, so this PR needs to be rebased on the updated topic-gpio branch.

nategraff-sifive commented:
Rebased on topic-gpio, still getting some spurious interrupts, still digging.

galak commented:
> Rebased on topic-gpio, still getting some spurious interrupts, still digging.

Any luck?  This is one of the few remaining GPIO drivers that needs to get converted.  Thanks.

pabigot commented:
I think that'd be appropriate.  We can continue to work on it after everything's merged to master.

nategraff-sifive committed:
drivers: gpio_sifive: Disable interrupts

The interrupt functionality of this driver is incomplete,
but for the sake of not slowing down the GPIO API refactor,
I'm just returning -ENOTSUP until we can track down the issue.

Signed-off-by: Nathaniel Graff <nathaniel.graff@sifive.com>

nategraff-sifive commented:
@galak, @pabigot, @mnkp `tests/drivers/gpio/gpio_api_1pin` passes now with the interrupt tests skipped

pabigot commented:
@nategraff-sifive Have you tried `tests/drivers/gpio/gpio_basic_api` (the 2-pin test)?

mnkp commented on drivers/gpio/gpio_sifive.c:
```diff
@@ -61,34 +61,80 @@ struct gpio_sifive_data {
 #define DEV_GPIO_DATA(dev)				\
 	((struct gpio_sifive_data *)(dev)->driver_data)
 
+/* _irq_level and _level2_irq are copied from
```
Let's prefix this paragraph with `TODO:`.

galak replied:
Added

mnkp reviewed (commented):
LGTM with a minor comment.

galak commented:
Going ahead and merging this and will open some issues for the TODO items.
